test/unit/test_block.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+import * as Helpers from './helpers.js';
+
+describe('block', () => {
+  it('block_untrusted', function () {
+    // Needs VIPS file load support
+    if (!Helpers.have('vipsload')) {
+      return this.skip();
+    }
+
+    // Some operations can be cached, which means they can return a result despite
+    // being blocked, if they were run previously. Disabling the cache fixes this issue.
+    Helpers.disableCache();
+
+    // vipsload is an untrusted operation, and should be blocked when blockUntrusted is true
+
+    vips.blockUntrusted(true);
+    expect(() => vips.Image.vipsload(Helpers.vipsFile)).to.throw(/operation is blocked/);
+
+    vips.blockUntrusted(false);
+    expect(() => vips.Image.vipsload(Helpers.vipsFile)).to.not.throw();
+
+    vips.blockUntrusted(true);
+    expect(() => vips.Image.vipsload(Helpers.vipsFile)).to.throw(/operation is blocked/);
+
+    // make sure no operations are blocked when the rest of the tests run
+    vips.blockUntrusted(false);
```

Comment:
This test is duplicated, so I think these lines can be removed.
```suggestion
```

Reply:
It depends on how thorough you want to test it. The idea is to test that it still blocks `vipsLoad` even after `vipsLoad` has been specifically greenlit and has been called. I felt like there could be muddy waters here, and I was _sort of, but not really_ right, because it actually doesn't pass this test when the cache is enabled.

Reply:
> It depends on how thorough you want to test it. The idea is to test that it still blocks `vipsLoad` even after `vipsLoad` has been specifically greenlit and has been called.

Ah, I see. Makes sense.

> it actually doesn't pass this test when the cache is enabled.

Oh curious, let me investigate that. I think the `operation is blocked`-error should always be thrown immediately and not be kept in the cache. Perhaps explicitly calling `.avg()` on the blocked operation would still cause that error?

/cc @jcupitt

Reply:
I already mentioned this for John, and he argued that since the `vipsLoad` method is never called, there is no problem serving the loaded image from cache.

I see it both ways. I think both approaches have upsides and downsides.

Reply:
Hi Kleis, I suppose when an operation is blocked we could search the cache and knock any operations of that type out. 

It wouldn't add security, since cached operations are not executed, but I suppose it might make behaviour less confusing. It's a cosmetic change heh.

I guess we could also check for blocked status on cache lookup, that might be simpler, and would avoid any threading issues.

Reply:
I misunderstood, sorry. I thought that blocked operations were cached, causing the error to be displayed only once, but this is not the case.

IIUC, the issue is that cached operations are not knocked out of the cache when `vips_block_untrusted_set( TRUE );` is called? I suppose this is a issue. For example, revisiting the test like this:
```js
// By default, libvips doesn't block untrusted operations
expect(() => vips.Image.vipsload(Helpers.vipsFile)).to.not.throw();

// However, if the environment variable `VIPS_BLOCK_UNTRUSTED` is set,
// or `vips_block_untrusted_set( TRUE );` is called, any operations
// that we've tagged as untrusted will be prevented from running
vips.blockUntrusted(true);

// For example, vipsload is an untrusted operation, and should throw
// when untrusted operations are blocked
// FIXME: Though, the first `vipsload` call will store the operation
// in cache, so subsequent calls will originate from their
expect(() => vips.Image.vipsload(Helpers.vipsFile)).to.throw(/operation is blocked/);

// Ensure no operations are blocked when the rest of the tests are run
vips.blockUntrusted(false);
```

Would cause a test failure.

> I guess we could also check for blocked status on cache lookup -- that might be simpler, and would avoid any threading issues.

Good idea. Another (simpler) idea is to just call `vips_cache_drop_all();` when `vips_block_untrusted_set( TRUE );` is called, to ensure that cached operations are invalidated.

Reply:
... though, that issue shouldn't block this PR, I just pushed commit ef0088e01814414bc6d6c2e10846952b7f7823af and will merge once the CI is green. :smile:

Reply:
I think it'd be simplest to add it here:

https://github.com/libvips/libvips/blob/master/libvips/iofuncs/cache.c#L801

Just swap that line for:

```C
                if( hit->invalid ||
                        (VIPS_OPERATION_GET_CLASS( hit->operation )->flags &
                         VIPS_OPERATION_BLOCKED) ) {
                        /* Has been tagged for removal, or has been blocked.
                         */
                        vips_cache_remove( hit->operation );
                        hit = NULL;
                }
```

It should be fast, and no race conditions are possible.

Reply:
Nice! Feel free to incorporate that, I think it would be useful for other bindings as well.

Reply:
... I confirm that this change works,  just pushed as commit 0fdd86bb820416562d9618a633e70f269174c98d, thanks!

Reply:
OK, pushed to master with https://github.com/libvips/libvips/commit/702ed8298f45d7ba342ebf5bae612d159e9cec6f

Reply:
Great! I think we're done, I'll mark this conversation as resolved.